### PR TITLE
Implement modular architecture for GoCardless transaction syncing

### PR DESCRIPTION
**Pull request by andrejvysny**
# Pull Request

Issue: #16 

## Description
This update introduces a robust, modular architecture for GoCardless transaction synchronization and related banking features. It adds new service, repository, and controller layers, comprehensive error handling, batch processing, and user-configurable sync options. The frontend is enhanced with new UI controls and hooks for sync settings and pagination, while backend APIs and models are extended to support these features.

```mermaid
sequenceDiagram
    participant User
    participant Frontend
    participant GoCardlessController
    participant GoCardlessService
    participant TokenManager
    participant GoCardlessBankData
    participant TransactionSyncService
    participant Repositories

    User->>Frontend: Initiate Sync (e.g. via UI button)
    Frontend->>GoCardlessController: POST /accounts/{id}/sync-transactions
    GoCardlessController->>GoCardlessService: syncAccountTransactions(accountId, user, options)
    GoCardlessService->>TokenManager: getAccessToken()
    TokenManager->>GoCardlessService: Return valid access token
    GoCardlessService->>GoCardlessBankData: getTransactions(accountId, dateFrom, dateTo)
    GoCardlessBankData->>GoCardlessService: Return transactions (with pagination, retries)
    GoCardlessService->>TransactionSyncService: syncTransactions(transactions, account, options)
    TransactionSyncService->>Repositories: Batch create/update transactions
    Repositories-->>TransactionSyncService: Sync results
    TransactionSyncService-->>GoCardlessService: Sync statistics
    GoCardlessService-->>GoCardlessController: Sync summary
    GoCardlessController-->>Frontend: JSON response (success/errors, stats)
    Frontend-->>User: Show sync results

```

Brief description of the changes made in this PR.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Security enhancement
- [ ] Refactoring (no functional changes)
- [ ] Test improvements
- [ ] CI/CD improvements

## Related Issues
Fixes #(issue number)
Relates to #(issue number)

## Changes Made
- [x] Backend changes (Laravel/PHP)
- [x] Frontend changes (React/TypeScript)
- [x] Database migrations/schema changes
- [x] API changes
- [ ] Configuration changes
- [ ] Documentation updates
- [ ] Test additions/modifications

### Detailed Changes
- Change 1: Description
- Change 2: Description
- Change 3: Description

## Financial Domain Impact
If this PR affects financial functionality:
- [ ] Transaction processing
- [ ] Account management
- [ ] Category/Tag/Merchant management
- [x] Bank integration (GoCardless)
- [ ] CSV import/export
- [ ] Financial calculations
- [ ] Currency handling
- [ ] Security/encryption
- [ ] Regulatory compliance

### Financial Data Accuracy
- [ ] All monetary calculations use appropriate decimal precision
- [ ] Currency handling is implemented correctly
- [ ] Rounding strategies are applied consistently
- [ ] No precision loss in financial operations

## Testing
- [ ] Tests pass locally
- [ ] New tests have been added for new functionality
- [ ] Manual testing completed
- [ ] Edge cases considered and tested
- [ ] Financial accuracy verified (if applicable)

### Test Coverage
- [ ] Unit tests
- [ ] Feature/Integration tests
- [ ] Browser tests (if UI changes)
- [ ] API tests (if backend changes)

## Security Considerations
- [ ] No sensitive data exposed in logs
- [ ] Input validation implemented
- [ ] Authentication/authorization maintained
- [ ] SQL injection prevention verified
- [ ] XSS protection maintained
- [ ] CSRF protection maintained
- [ ] Financial data encryption verified (if applicable)

## Breaking Changes
If this PR introduces breaking changes, please describe:
- What breaks?
- Migration path for users
- Documentation updates needed

## Screenshots/Videos
If applicable, add screenshots or videos to demonstrate the changes.

**Before:**
[Screenshot/Video of before state]

**After:**
[Screenshot/Video of after state]

## Database Changes
- [x] Database migrations included
- [x] Migrations are reversible
- [ ] Data integrity maintained
- [ ] Performance impact considered
- [ ] Indexes added where necessary

## API Changes
If this PR modifies APIs:
- [x] API documentation updated
- [ ] Backward compatibility maintained or breaking changes documented
- [ ] Rate limiting considered
- [ ] Error handling improved
- [ ] OpenAPI/Swagger specs updated

## Performance Impact
- [x] No significant performance degradation
- [ ] Database queries optimized
- [ ] Frontend bundle size impact minimal
- [ ] Caching strategies implemented where beneficial
- [ ] Memory usage optimized

## Documentation
- [x] Code is self-documenting
- [ ] PHPDoc blocks added/updated
- [ ] TypeScript types properly defined
- [ ] README updated (if needed)
- [ ] API documentation updated (if needed)
- [ ] User documentation updated (if needed)

## Deployment Considerations
- [ ] Environment variables documented (if new ones added)
- [ ] Docker configuration updated (if needed)
- [ ] Deployment scripts updated (if needed)
- [ ] No manual intervention required for deployment
- [ ] Rollback plan considered

## Checklist
- [ ] My code follows the project's style guidelines
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

## Additional Notes
Any additional information that reviewers should know about this PR. 

## Summary by CodeRabbit

- **New Features**
  - Added configurable sync options for accounts, allowing users to set preferences for updating existing transactions and forcing full syncs (up to 90 days).
  - Introduced UI controls for managing sync options and integrated them into the transaction sync process.
  - Added endpoints to sync transactions for individual or all accounts with new sync options.
  - Implemented paginated transaction loading with a new "load more" experience and visible transaction counts.
  - Introduced a custom hook for managing paginated data fetching and state.
  - Introduced comprehensive GoCardless integration with modular services for token management, transaction syncing, account import, and bank data handling.
  - Added detailed error handling and logging for GoCardless API interactions.
  - Added extensive new architecture and analysis documentation for CSV services, GoCardless integration, and Firefly III systems.
  - Added new testing infrastructure including Docker-based test environment, coverage reporting, and extensive unit tests for GoCardless components.
  - Added new API routes and UI support for managing GoCardless account sync options.
  - Added new service providers and repositories to support GoCardless integration.
  - Added new factory for Account model to support testing.
  - Introduced a new React hook for load-more pagination and refactored transaction page to use it.
  - Added detailed transaction filtering schemas for improved validation and filtering.

- **Improvements**
  - Refactored transaction filtering, pagination, and data fetching logic for better maintainability and user experience.
  - Enhanced error handling and reliability in bank data synchronization.
  - Centralized and modularized transaction syncing, token management, and data mapping for GoCardless integration.
  - Standardized API responses and improved performance for bulk transaction operations.
  - Improved type safety and explicit return types in multiple controllers.
  - Middleware and request handling now conditionally access facade-dependent data only when the application is fully booted.
  - Refactored transaction rule pipeline to correctly persist processed transactions.
  - Added factory support for Account model and improved model casting for sync options.
  - Improved Docker environment with SQLite support and optional Xdebug configuration.
  - Updated service provider registration to include new GoCardless services and centralized route management.
  - Enhanced GoCardless client with pagination, caching, and retry mechanisms.
  - Updated Docker Compose configuration and added scripts for development and testing workflows.
  - Improved frontend components with refined type usage and UI enhancements.
  - Added authorization checks and improved return type declarations in multiple controllers.
  - Improved logging and error reporting in GoCardless controllers and services.
  - Added detailed date range calculation logic for transaction syncs.

- **Bug Fixes**
  - Restored duplicate transaction detection during imports.
  - Fixed transaction rule pipeline to persist processed transactions correctly.

- **Chores**
  - Updated service provider registration and configuration for better dependency management.
  - Adjusted Docker Compose and workflow trigger settings.
  - Added database migration to support new sync options on accounts.
  - Added new routes and route names for GoCardless sync endpoints and account sync options update.
  - Added authorization checks and improved return type declarations in multiple controllers.
  - Added new scripts for Docker-based development and testing workflows.
  - Updated .gitignore to exclude coverage and PHPUnit cache directories.

### CODE REVIEW

File: app/Repositories/TransactionRepository.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        $count = 0;
+        
+        DB::transaction(function () use ($updates, &$count) {
+            foreach ($updates as $transactionId => $data) {
+                $updated = Transaction::where('transaction_id', $transactionId)
+                    ->update($data);
+                if ($updated) {
+                    $count++;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (andrejvysny) [reply app/Repositories/TransactionRepository.php:84]: This will be implemented in different Issue: #20
